File: backend/java/src/main/java/com/ragassistant/provider/embedding/LMStudioEmbeddingProvider.java
package com.ragassistant.provider.embedding;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ragassistant.config.IngestionConfig;
import com.ragassistant.model.DocumentChunk;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class LMStudioEmbeddingProvider implements EmbeddingProvider {
    private final IngestionConfig config;
    private final OkHttpClient httpClient = new OkHttpClient();
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public List<DocumentChunk> embed(List<DocumentChunk> chunks) {
        for (DocumentChunk chunk : chunks) {
            log.info("[LM Studio Embed] Starting embedding for chunk: {}", chunk.getId());

            try {
                String payload = objectMapper.writeValueAsString(Map.of(
                        "model", config.getEmbeddingModel(),
                        "input", chunk.getText()));

                RequestBody body = RequestBody.create(
                        payload,
                        MediaType.parse("application/json; charset=utf-8"));

                Request request = new Request.Builder()
                        .url(config.getLmStudioApiUrl() + "/embeddings")
                        .post(body)
                        .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        log.error("[LM Studio Embed Error] HTTP Status: {}", response.code());
                        continue;
                    }

                    JsonNode root = objectMapper.readTree(response.body().string());
                    JsonNode embeddingNode = root.path("data").get(0).path("embedding");

                    List<Double> vector = new ArrayList<>();
                    if (embeddingNode.isArray()) {
                        for (JsonNode val : embeddingNode) {
                            vector.add(val.asDouble());
                        }
                    }

                    chunk.getMetadata().setEmbedding(vector);
                    log.info("[LM Studio Embed] Success. Vector length: {}", vector.size());
                }
            } catch (IOException e) {
                log.error("[LM Studio Embed Error] {}", e.getMessage());
            }
        }
        return chunks;
    }
}

File: backend/javascript/src/services/providers/embeddings/lmstudio-embedding.provider.ts
import { Injectable, Logger } from "@nestjs/common";
import { EmbeddingProvider } from "./embedding.provider";
import { IngestionConfig } from "src/services/config/ingestion.config";
import OpenAI from "openai";
import { DocumentChunk } from "src/services/models/document-chunk";

@Injectable()
export class LMStudioEmbeddingProvider extends EmbeddingProvider {
    private readonly logger = new Logger(LMStudioEmbeddingProvider.name);
    private readonly client: OpenAI;

    constructor(private readonly config: IngestionConfig) {
        super();
        this.client = new OpenAI({
            apiKey: 'lm-studio-api-key', // 로컬이라 api key 상관 없음
            baseURL: this.config.lmStudioApiUrl,
        });
    }

    async embed(chunks: DocumentChunk[]): Promise<DocumentChunk[]> {
        for (const chunk of chunks) {

            this.logger.log(`[LM Studio Embed] Starting embedding for ${chunks.length} chunks...`);
            this.logger.log(`[LM Studio Embed] Using model: ${this.config.embeddingModel}`);

            try {
                const response = await fetch(`${this.config.lmStudioApiUrl}/embeddings`, {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({
                        model: this.config.embeddingModel,
                        input: chunk.text
                    })
                });

                if (!response.ok) {
                    throw new Error(`LM Studio API Error: ${response.statusText}`);
                }

                const data: any = await response.json();
                const embedding = data.data[0].embedding;
                chunk.metadata.embedding = embedding;

            } catch (err: any) {
                this.logger.error(`[LM Studio Embed Error] Chunk ID: ${chunk.id}`);
                this.logger.error(`Error Message: ${err.message}`);

                // 만약 API 응답 자체가 있다면 그것도 출력
                if (err.response) {
                    this.logger.error(`Response Data: ${JSON.stringify(err.response.data)}`);
                }
            }
        }
        return chunks;
    }
}